### ISSUE TITLE
OpApplyDirectStyling created on text replacement fails in collab mode

### ISSUE THREAD

**Issue by kossebau** [kogmbh/WebODF#551]
Each time when overwriting existing text with other text also a OpApplyDirectStyling is created with the complete style properties of the first character in the selection, to ensure the current styling is kept (at least that way reasoned in #296, commit e2b7b378f152dc976836537175f18e8678a6e243).

While this works fine in single-editor mode, because the styles are resolved to nothing when executed by odf.TextStyleApplicator in the same style environment, it is a problem in collab mode, because it is handled as a serious op, with the intention to style with all the properties, which only results in unexpected stylings, as we saw today.

No idea yet how to apporach the original problem in another, less side-effecting way.

**Comment by vandenoever** [comment]
What I fail to see in the given example, "replace 'about' with 'o'" is why a style is applied at all? I'd think that the intent would be for the use to give 'o' the same style as 'about'. By applying InsertText first and then RemoveText, the style would be kept.

**Comment by peitschie** [comment]
> By applying InsertText first and then RemoveText, the style would be kept.

That's the user experience we've had to implement, yes :wink: 

The problem is, the step system does not allow text insertion inside the (now) empty span, so in order to achieve this user behaviour, we re-apply the style after text insertion.

**Comment by vandenoever** [comment]
How can the span be empty?
```
'about' -> InsertText 'o' -> 'abouto' -> RemoveText 'about' -> 'o'
```
If the controller creates the operations like that, then at least the issues does not occur in the current UI. I'm asking because OwnCloud would like to update to a newer WebODF. If issue #551 does not occur in practice, they could upgrade.

**Comment by peitschie** [comment]
Agreed that the described scenario would avoid the issue (at least, for this specific case, haven't considered hard enough whether it has it's own version of this bug).

Unfortunately, that is not what the current code does. The existing code does the remove first, followed by the insert. So, in the existing implementation, the span is undoubtedly empty.

This is only one of the difficulties however, as similar problems can still be created in collab mode no matter how you do the insert/remove pair.

E.g., client A does:
``'about' -> InsertText 'o' -> 'abouto'``

client B does:
``select 'about' -> RemoveText``

When the insertion eventually arrives at client B, the span was already long destroyed that client A inserted it into.

**Comment by peitschie** [comment]
> I'm asking because OwnCloud would like to update to a newer WebODF. If issue #551 does not occur in practice, they could upgrade.

Which version is OwnCloud currently using? Does their current version have collab mode?

**Comment by vandenoever** [comment]
They use collaborative editing with webodf 0.4.2
https://github.com/owncloud/documents/blob/master/js/3rdparty/webodf/webodf.js
and have some serious issues trying to upgrade:
https://github.com/owncloud/documents/pull/419

**Comment by peitschie** [comment]
Based on the comments on that issue, that is different possibly from this current one. This issue is about a specific edge case when working in collab mode which as far as I can tell has existed as long as we've attempted to support direct formatting in collab.

Are we able to get more specific reproduction cases from them?

**Comment by peitschie** [comment]
Just for record keeping, some discussion on this has already occurred in https://github.com/kogmbh/WebODF/issues/37#issuecomment-25418471

**Comment by VicDeo** [comment]
@peitschie exact reproduction steps are listed in https://github.com/owncloud/documents/pull/354#issuecomment-54651167

**Comment by peitschie** [comment]
> @peitschie exact reproduction steps are listed in owncloud/documents#354 (comment)
> 
> UPD: I mean the first and the second issues with applying font size and bold/italic/etc, the third issue is surely unrelated to WebODF.

Are you able to drop by IRC and discuss those issues sometime? They appear unrelated to the bug described here. The font sizes & bold/italics etc. are all currently working in trunk's split-screen editor demo.

**Comment by VicDeo** [comment]
@peitschie yes, I've tried to sync from scratch and it does work now. Strange.

**Comment by peitschie** [comment]
> @peitschie yes, I've tried to sync from scratch and it does work now. Strange.

Indeed, that is unexpected. Still... glad to hear it works :smile:
